File: src/components/survival/phrases/SurvivalCompositePhraseEngine.ts
/**
 * Survival composite phrase boss: parallel candidate filtering then single phrase (no chord loop).
 */
import type {
  SurvivalPhraseChord,
  SurvivalPhraseDefinition,
} from '@/utils/survivalPhraseDefinitions';

export type SurvivalCompositePhraseNoteResult =
  | 'progress'
  | 'measure-complete'
  | 'phrase-complete'
  | 'miss';

export interface SurvivalCompositePhraseCandidateState {
  readonly sourceStageNumber: number;
  readonly phrase: SurvivalPhraseDefinition;
  readonly chordIndex: number;
  readonly targetNoteIndex: number;
  readonly correctNoteIndices: ReadonlySet<number>;
  readonly revealedNoteIndices: ReadonlySet<number>;
}

export interface SurvivalCompositePhraseRuntimeState {
  readonly sourcePhrases: readonly SurvivalPhraseDefinition[];
  readonly candidates: readonly SurvivalCompositePhraseCandidateState[];
  readonly lockedSourceStageNumber: number | null;
  readonly lastCompletedSourceStageNumber: number | null;
}

function candidateFromPhrase(phrase: SurvivalPhraseDefinition): SurvivalCompositePhraseCandidateState {
  return {
    sourceStageNumber: phrase.stageNumber,
    phrase,
    chordIndex: 0,
    targetNoteIndex: 0,
    correctNoteIndices: new Set(),
    revealedNoteIndices: new Set(),
  };
}

export function createInitialCompositePhraseRuntimeState(
  sourcePhrases: readonly SurvivalPhraseDefinition[],
): SurvivalCompositePhraseRuntimeState {
  return {
    sourcePhrases,
    candidates: sourcePhrases.map(candidateFromPhrase),
    lockedSourceStageNumber: null,
    lastCompletedSourceStageNumber: null,
  };
}

function getCurrentChord(
  c: SurvivalCompositePhraseCandidateState,
): SurvivalPhraseChord | null {
  return c.phrase.chords[c.chordIndex] ?? null;
}

function getTargetNote(c: SurvivalCompositePhraseCandidateState) {
  const chord = getCurrentChord(c);
  if (!chord) return null;
  return chord.notes[c.targetNoteIndex] ?? null;
}

function isChordComplete(chord: SurvivalPhraseChord, correctIndices: ReadonlySet<number>): boolean {
  return chord.notes.length > 0 && correctIndices.size >= chord.notes.length;
}

function isLastChord(c: SurvivalCompositePhraseCandidateState): boolean {
  return c.chordIndex >= c.phrase.chords.length - 1;
}

function resetChordFields(
  c: SurvivalCompositePhraseCandidateState,
): SurvivalCompositePhraseCandidateState {
  return {
    ...c,
    targetNoteIndex: 0,
    correctNoteIndices: new Set(),
    revealedNoteIndices: new Set(),
  };
}

/** 次コードへ（ループなし; 呼び出し元で最終小節を弾く） */
function advanceToNextChord(
  c: SurvivalCompositePhraseCandidateState,
): SurvivalCompositePhraseCandidateState {
  const nextChordIndex = c.chordIndex + 1;
  return {
    ...c,
    chordIndex: nextChordIndex,
    targetNoteIndex: 0,
    correctNoteIndices: new Set(),
    revealedNoteIndices: new Set(),
  };
}

export interface SurvivalCompositePhraseNoteEvaluation {
  readonly result: SurvivalCompositePhraseNoteResult;
  readonly nextState: SurvivalCompositePhraseRuntimeState;
}

function resetAllCandidates(
  state: SurvivalCompositePhraseRuntimeState,
  preserveLastCompleted: boolean,
): SurvivalCompositePhraseRuntimeState {
  return {
    sourcePhrases: state.sourcePhrases,
    candidates: state.sourcePhrases.map(candidateFromPhrase),
    lockedSourceStageNumber: null,
    lastCompletedSourceStageNumber: preserveLastCompleted
      ? state.lastCompletedSourceStageNumber
      : null,
  };
}

function applyKnownGoodStep(
  c: SurvivalCompositePhraseCandidateState,
  pitchClass: number,
): { candidate: SurvivalCompositePhraseCandidateState; result: SurvivalCompositePhraseNoteResult } {
  const chord = getCurrentChord(c);
  const target = getTargetNote(c);
  if (!chord || !target) {
    return { candidate: c, result: 'miss' };
  }
  const allowed = new Set(chord.notes.map((n) => n.pitchClass));
  if (!allowed.has(pitchClass) || pitchClass !== target.pitchClass) {
    return { candidate: resetChordFields(c), result: 'miss' };
  }
  const nextCorrect = new Set(c.correctNoteIndices);
  nextCorrect.add(c.targetNoteIndex);
  const nextRevealed = new Set(c.revealedNoteIndices);
  nextRevealed.add(c.targetNoteIndex);

  if (isChordComplete(chord, nextCorrect)) {
    if (isLastChord(c)) {
      return {
        candidate: {
          ...c,
          correctNoteIndices: nextCorrect,
          revealedNoteIndices: nextRevealed,
        },
        result: 'phrase-complete',
      };
    }
    return {
      candidate: advanceToNextChord({
        ...c,
        correctNoteIndices: nextCorrect,
        revealedNoteIndices: nextRevealed,
      }),
      result: 'measure-complete',
    };
  }

  return {
    candidate: {
      ...c,
      targetNoteIndex: c.targetNoteIndex + 1,
      correctNoteIndices: nextCorrect,
      revealedNoteIndices: nextRevealed,
    },
    result: 'progress',
  };
}

function selectionMatchingIndices(
  state: SurvivalCompositePhraseRuntimeState,
  pitchClass: number,
): number[] {
  const out: number[] = [];
  for (let i = 0; i < state.candidates.length; i += 1) {
    const note = getTargetNote(state.candidates[i]);
    if (note && note.pitchClass === pitchClass) {
      out.push(i);
    }
  }
  return out;
}

function replaceCandidateAt(
  candidates: readonly SurvivalCompositePhraseCandidateState[],
  index: number,
  next: SurvivalCompositePhraseCandidateState,
): SurvivalCompositePhraseCandidateState[] {
  const copy = [...candidates];
  copy[index] = next;
  return copy;
}

export function evaluateCompositePhraseNoteOn(
  state: SurvivalCompositePhraseRuntimeState,
  pitchClass: number,
): SurvivalCompositePhraseNoteEvaluation {
  const pc = ((pitchClass % 12) + 12) % 12;

  if (state.lockedSourceStageNumber !== null) {
    const idx = state.candidates.findIndex(
      (c) => c.sourceStageNumber === state.lockedSourceStageNumber,
    );
    if (idx < 0) {
      return { result: 'miss', nextState: resetAllCandidates(state, true) };
    }
    const cur = state.candidates[idx];
    const { candidate, result } = applyKnownGoodStep(cur, pc);
    if (result === 'miss') {
      return { result: 'miss', nextState: resetAllCandidates(state, true) };
    }
    if (result === 'phrase-complete') {
      const finishedStage = candidate.sourceStageNumber;
      return {
        result: 'phrase-complete',
        nextState: resetAllCandidates(
          {
            ...state,
            lastCompletedSourceStageNumber: finishedStage,
          },
          true,
        ),
      };
    }
    return {
      result,
      nextState: {
        ...state,
        candidates: replaceCandidateAt(state.candidates, idx, candidate),
        lockedSourceStageNumber: candidate.sourceStageNumber,
      },
    };
  }

  const matchIdx = selectionMatchingIndices(state, pc);
  if (matchIdx.length === 0) {
    return { result: 'miss', nextState: resetAllCandidates(state, true) };
  }

  const advanced: SurvivalCompositePhraseCandidateState[] = [];
  let aggResult: SurvivalCompositePhraseNoteResult = 'progress';
  for (const i of matchIdx) {
    const { candidate, result } = applyKnownGoodStep(state.candidates[i], pc);
    if (result === 'miss') {
      return { result: 'miss', nextState: resetAllCandidates(state, true) };
    }
    if (result === 'phrase-complete') {
      const finishedStage = candidate.sourceStageNumber;
      return {
        result: 'phrase-complete',
        nextState: resetAllCandidates(
          {
            ...state,
            lastCompletedSourceStageNumber: finishedStage,
          },
          true,
        ),
      };
    }
    advanced.push(candidate);
    aggResult = result;
  }

  const lock = advanced.length === 1 ? advanced[0].sourceStageNumber : null;
  return {
    result: aggResult,
    nextState: {
      ...state,
      candidates: advanced,
      lockedSourceStageNumber: lock,
    },
  };
}

/** 並行選択中: 全候補の現小節で、順位ごとに同ピッチクラスかつ revealed 済みの共通 prefix 長さ */
export function compositeSelectionGreenPrefixLength(
  state: SurvivalCompositePhraseRuntimeState,
): number {
  if (state.lockedSourceStageNumber !== null) return 0;
  if (state.candidates.length === 0) return 0;
  const chordLens = state.candidates.map((c) => getCurrentChord(c)?.notes.length ?? 0);
  const maxSafe = Math.min(...chordLens);
  let p = 0;
  outer: while (p < maxSafe) {
    let pitchClassBaseline: number | null = null;
    for (const c of state.candidates) {
      const ch = getCurrentChord(c);
      const noteAt = ch?.notes[p];
      if (!noteAt || !c.revealedNoteIndices.has(p)) {
        break outer;
      }
      if (pitchClassBaseline === null) {
        pitchClassBaseline = noteAt.pitchClass;
      } else if (noteAt.pitchClass !== pitchClassBaseline) {
        break outer;
      }
    }
    p += 1;
  }
  return p;
}

export interface SurvivalCompositePhraseStaffChordView {
  readonly chord: SurvivalPhraseChord | null;
  readonly correctNoteIndices: ReadonlySet<number>;
}

export function getCompositePhraseStaffChordView(
  state: SurvivalCompositePhraseRuntimeState,
): SurvivalCompositePhraseStaffChordView {
  if (state.candidates.length === 0) {
    return { chord: null, correctNoteIndices: new Set() };
  }
  if (state.lockedSourceStageNumber !== null) {
    const c = state.candidates.find((x) => x.sourceStageNumber === state.lockedSourceStageNumber);
    if (!c) {
      return { chord: null, correctNoteIndices: new Set() };
    }
    return {
      chord: getCurrentChord(c),
      correctNoteIndices: c.revealedNoteIndices,
    };
  }

  const template = state.candidates[0];
  const chord = getCurrentChord(template);
  const len = compositeSelectionGreenPrefixLength(state);
  const correct = new Set<number>();
  for (let i = 0; i < len; i += 1) {
    correct.add(i);
  }
  return {
    chord,
    correctNoteIndices: correct,
  };
}

File: ios/Jazzify/Survival/Phrase/SurvivalCompositePhraseEngine.swift
import Foundation

/// Web `compositePhraseEngine.ts` の Swift 移植。全候補 KMP 並列判定（コードループなし）。
enum SurvivalCompositePhraseNoteResult: Equatable {
    case progress
    case resync
    case measureComplete
    case phraseComplete
    case miss
}

struct SurvivalCompositePhraseCandidateState: Equatable {
    let sourceStageNumber: Int
    let phrase: SurvivalPhraseDefinition
    var chordIndex: Int
    var targetNoteIndex: Int
    var correctNoteIndices: Set<Int>
    var revealedNoteIndices: Set<Int>
}

struct SurvivalCompositePhraseRuntimeState: Equatable {
    let sourcePhrases: [SurvivalPhraseDefinition]
    var candidates: [SurvivalCompositePhraseCandidateState]
    var primarySourceStageNumber: Int?
    /// 直近でフレーズ完成したソースステージ番号（重複完成ダメージ用）。
    var lastCompletedSourceStageNumber: Int?
}

struct SurvivalCompositePhraseStaffChordView: Equatable {
    let chord: SurvivalPhraseChord?
    let correctNoteIndices: Set<Int>
}

enum SurvivalCompositePhraseEngine {
    static func createInitialState(sourcePhrases: [SurvivalPhraseDefinition]) -> SurvivalCompositePhraseRuntimeState {
        SurvivalCompositePhraseRuntimeState(
            sourcePhrases: sourcePhrases,
            candidates: sourcePhrases.map(candidateFromPhrase),
            primarySourceStageNumber: nil,
            lastCompletedSourceStageNumber: nil
        )
    }

    static func evaluateNoteOn(
        state: SurvivalCompositePhraseRuntimeState,
        pitchClass rawPc: Int
    ) -> (result: SurvivalCompositePhraseNoteResult, nextState: SurvivalCompositePhraseRuntimeState) {
        let pc = PhraseStreamMatching.normalizedPitchClass(rawPc)

        var candidateByStage: [Int: SurvivalCompositePhraseCandidateState] = [:]
        for c in state.candidates {
            candidateByStage[c.sourceStageNumber] = c
        }

        let steps = state.sourcePhrases.map { phrase in
            let current = candidateByStage[phrase.stageNumber] ?? candidateFromPhrase(phrase)
            return applyStreamingCandidateStep(candidate: current, pitchClass: pc)
        }

        let completed = steps.filter { $0.result == .phraseComplete }
        if !completed.isEmpty {
            let preferred = state.primarySourceStageNumber.flatMap { primary in
                completed.first(where: { $0.candidate.sourceStageNumber == primary })
            }
            let finished = preferred ?? completed[0]
            var st = state
            st.lastCompletedSourceStageNumber = finished.candidate.sourceStageNumber
            st.candidates = steps.map(\.candidate)
            return (.phraseComplete, resetAllCandidates(st, preserveLastCompleted: true))
        }

        let accepted = steps.filter { $0.accepted && $0.matchedLength > 0 }
        if accepted.isEmpty {
            var st = state
            st.candidates = steps.map(\.candidate)
            return (.miss, resetAllCandidates(st, preserveLastCompleted: true))
        }

        let bestMatchedLength = accepted.map(\.matchedLength).max() ?? 0
        let nextPrimary = resolvePrimaryStageNumber(
            state: state,
            steps: steps,
            bestMatchedLength: bestMatchedLength
        )

        let selectedStep: CandidateStep? = {
            if let primary = nextPrimary {
                return steps.first(where: { $0.candidate.sourceStageNumber == primary })
            }
            return accepted.first(where: { $0.matchedLength == bestMatchedLength })
        }()

        let primaryResync =
            selectedStep.map { $0.matchedLength < $0.beforeMatchedLength } ?? false
        let result = resolveAggregateResult(selectedStep: selectedStep, primaryResync: primaryResync)

        var next = state
        next.candidates = steps.map(\.candidate)
        next.primarySourceStageNumber = nextPrimary
        return (result, next)
    }

    static func selectionGreenPrefixLength(state: SurvivalCompositePhraseRuntimeState) -> Int {
        if state.primarySourceStageNumber != nil { return 0 }

        let candidates = compositeSelectionCandidates(state: state)
        guard !candidates.isEmpty else { return 0 }

        let chordLens = candidates.map { c in
            chord(at: c.chordIndex, phrase: c.phrase)?.notes.count ?? 0
        }
        guard let maxSafe = chordLens.min(), maxSafe > 0 else { return 0 }

        var p = 0
        while p < maxSafe {
            var baseline: Int?
            for c in candidates {
                guard let ch = chord(at: c.chordIndex, phrase: c.phrase),
                      let noteAt = ch.notes[safe: p],
                      c.revealedNoteIndices.contains(p)
                else {
                    return p
                }
                if let b = baseline, b != noteAt.pitchClass {
                    return p
                }
                if baseline == nil {
                    baseline = noteAt.pitchClass
                }
            }
            p += 1
        }
        return p
    }

    static func staffChordView(state: SurvivalCompositePhraseRuntimeState) -> SurvivalCompositePhraseStaffChordView {
        guard !state.candidates.isEmpty else {
            return SurvivalCompositePhraseStaffChordView(chord: nil, correctNoteIndices: [])
        }

        if let primary = state.primarySourceStageNumber {
            guard let c = state.candidates.first(where: { $0.sourceStageNumber == primary }) else {
                return SurvivalCompositePhraseStaffChordView(chord: nil, correctNoteIndices: [])
            }
            return SurvivalCompositePhraseStaffChordView(
                chord: chord(at: c.chordIndex, phrase: c.phrase),
                correctNoteIndices: c.revealedNoteIndices
            )
        }

        guard let display = getDisplayCandidate(state: state) else {
            return SurvivalCompositePhraseStaffChordView(chord: nil, correctNoteIndices: [])
        }

        let len = selectionGreenPrefixLength(state: state)
        return SurvivalCompositePhraseStaffChordView(
            chord: chord(at: display.chordIndex, phrase: display.phrase),
            correctNoteIndices: PhraseStreamMatching.prefixIndexSet(len)
        )
    }

    // MARK: - Internal

    private struct CandidateStep {
        let candidate: SurvivalCompositePhraseCandidateState
        let result: SurvivalCompositePhraseNoteResult
        let accepted: Bool
        let matchedLength: Int
        let beforeMatchedLength: Int
    }

    private static func candidateFromPhrase(_ phrase: SurvivalPhraseDefinition) -> SurvivalCompositePhraseCandidateState {
        SurvivalCompositePhraseCandidateState(
            sourceStageNumber: phrase.stageNumber,
            phrase: phrase,
            chordIndex: 0,
            targetNoteIndex: 0,
            correctNoteIndices: [],
            revealedNoteIndices: []
        )
    }

    private static func chord(at index: Int, phrase: SurvivalPhraseDefinition) -> SurvivalPhraseChord? {
        phrase.chords[safe: index]
    }

    private static func candidateWithMatchedLength(
        _ c: SurvivalCompositePhraseCandidateState,
        matchedLength: Int
    ) -> SurvivalCompositePhraseCandidateState {
        let coord = PhraseStreamMatching.coordinateFromMatchedLength(
            chords: c.phrase.chords,
            matchedLength: matchedLength
        )
        let correct = PhraseStreamMatching.prefixIndexSet(coord.targetNoteIndex)
        var next = c
        next.chordIndex = coord.chordIndex
        next.targetNoteIndex = coord.targetNoteIndex
        next.correctNoteIndices = correct
        next.revealedNoteIndices = correct
        return next
    }

    private static func applyStreamingCandidateStep(
        candidate c: SurvivalCompositePhraseCandidateState,
        pitchClass pc: Int
    ) -> CandidateStep {
        let cache = PhraseStreamMatching.getCompositeKmpCache(chords: c.phrase.chords)
        let pattern = cache.pattern

        if pattern.isEmpty {
            return CandidateStep(
                candidate: candidateFromPhrase(c.phrase),
                result: .miss,
                accepted: false,
                matchedLength: 0,
                beforeMatchedLength: 0
            )
        }

        let beforeMatchedLength = PhraseStreamMatching.matchedLengthFromCoordinates(
            chords: c.phrase.chords,
            chordIndex: c.chordIndex,
            targetNoteIndex: c.targetNoteIndex
        )
        let nextMatchedLength = PhraseStreamMatching.advanceKmp(
            pattern: cache.pattern,
            table: cache.table,
            matchedLength: beforeMatchedLength,
            pitchClass: pc
        )

        if nextMatchedLength == 0 {
            return CandidateStep(
                candidate: candidateFromPhrase(c.phrase),
                result: .miss,
                accepted: false,
                matchedLength: 0,
                beforeMatchedLength: beforeMatchedLength
            )
        }

        let nextCandidate = candidateWithMatchedLength(c, matchedLength: nextMatchedLength)

        if nextMatchedLength >= pattern.count {
            return CandidateStep(
                candidate: nextCandidate,
                result: .phraseComplete,
                accepted: true,
                matchedLength: nextMatchedLength,
                beforeMatchedLength: beforeMatchedLength
            )
        }

        if PhraseStreamMatching.isNonFinalMeasureBoundary(
            chords: c.phrase.chords,
            matchedLength: nextMatchedLength
        ) {
            return CandidateStep(
                candidate: nextCandidate,
                result: .measureComplete,
                accepted: true,
                matchedLength: nextMatchedLength,
                beforeMatchedLength: beforeMatchedLength
            )
        }

        return CandidateStep(
            candidate: nextCandidate,
            result: .progress,
            accepted: true,
            matchedLength: nextMatchedLength,
            beforeMatchedLength: beforeMatchedLength
        )
    }

    private static func compositeSelectionCandidates(
        state: SurvivalCompositePhraseRuntimeState
    ) -> [SurvivalCompositePhraseCandidateState] {
        if let primary = state.primarySourceStageNumber {
            if let c = state.candidates.first(where: { $0.sourceStageNumber == primary }) {
                return [c]
            }
            return []
        }

        guard !state.candidates.isEmpty else { return [] }

        var best = 0
        for c in state.candidates {
            best = max(
                best,
                PhraseStreamMatching.matchedLengthFromCoordinates(
                    chords: c.phrase.chords,
                    chordIndex: c.chordIndex,
                    targetNoteIndex: c.targetNoteIndex
                )
            )
        }

        if best <= 0 {
            return state.candidates
        }

        return state.candidates.filter { c in
            PhraseStreamMatching.matchedLengthFromCoordinates(
                chords: c.phrase.chords,
                chordIndex: c.chordIndex,
                targetNoteIndex: c.targetNoteIndex
            ) == best
        }
    }

    private static func getDisplayCandidate(
        state: SurvivalCompositePhraseRuntimeState
    ) -> SurvivalCompositePhraseCandidateState? {
        compositeSelectionCandidates(state: state).first
    }

    private static func resolvePrimaryStageNumber(
        state: SurvivalCompositePhraseRuntimeState,
        steps: [CandidateStep],
        bestMatchedLength: Int
    ) -> Int? {
        let best = steps.filter { $0.accepted && $0.matchedLength == bestMatchedLength }

        if let previousPrimary = state.primarySourceStageNumber,
           let previousStep = steps.first(where: { $0.candidate.sourceStageNumber == previousPrimary }),
           previousStep.accepted,
           previousStep.matchedLength == bestMatchedLength {
            return previousPrimary
        }

        if best.count == 1 {
            return best[0].candidate.sourceStageNumber
        }

        return nil
    }

    private static func resolveAggregateResult(
        selectedStep: CandidateStep?,
        primaryResync: Bool
    ) -> SurvivalCompositePhraseNoteResult {
        if primaryResync {
            return .resync
        }
        if selectedStep?.result == .measureComplete {
            return .measureComplete
        }
        return .progress
    }

    private static func resetAllCandidates(
        _ state: SurvivalCompositePhraseRuntimeState,
        preserveLastCompleted: Bool
    ) -> SurvivalCompositePhraseRuntimeState {
        var next = state
        next.candidates = state.sourcePhrases.map(candidateFromPhrase)
        next.primarySourceStageNumber = nil
        if !preserveLastCompleted {
            next.lastCompletedSourceStageNumber = nil
        }
        return next
    }
}

private extension Array {
    subscript(safe index: Int) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}
